---
title: "JB6 Mouse Cells Methy-seq"
author: Davit; Komal
date: 2/4/2022
version: 1.0
output:
  html_notebook:
    highlight: tango
    toc: true
    toc_float: true
    toc_depth: 6
    number_sections: yes
    code_folding: hide
---

Data location: [](https://rutgers.app.box.com/folder/118765606470?s=m5p30wnhsc1rl3mooezuj4odu9yu2jkh)

```{r}
# if (!requireNamespace("BiocManager",
#                       quietly = TRUE))
#   install.packages("BiocManager")
# 
# BiocManager::install("ChIPseeker")
# BiocManager::install("GO.db")
# BiocManager::install("DO.db")

# BiocManager::install("org.Mm.eg.db")
# BiocManager::install("TxDb.Mmusculus.UCSC.mm10.knownGene")

# BiocManager::install("org.Hs.eg.db")
# BiocManager::install("TxDb.Hsapiens.UCSC.hg38.knownGene")

# BiocManager::install("DSS")
```
# Resources
1. [ChiPseeker package](http://bioconductor.org/packages/devel/bioc/vignettes/ChIPseeker/inst/doc/ChIPseeker.html)  
  
2. [DSS package](https://bioconductor.org/packages/release/bioc/vignettes/DSS/inst/doc/DSS.html)  
  
3. [Differential methylation analysis for bisulfite sequencing using DSS](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8171293/)  
  
# Introduction
ChiP-seq is a method to analyze protein interaction with DNA and combines chromatin immunoprecipitation (CHIP) with massively parallel DNA sequencing to identify binding sites
  
# Setup
```{r setup}
require(data.table)
require(ggplot2)
require(ChIPseeker)
require(TxDb.Mmusculus.UCSC.mm10.knownGene)
require(org.Mm.eg.db)

require(DSS)
require(bsseq)
```

# Load data
```{r}
dt1 <- fread("data/combined_def.tsv")
colnames(dt1)
```

```{r}
setorder(dt1,
         start)
dt1$ndx <- 1:nrow(dt1)
```

```{r}
write.table(dt1[, c("chr",
                    "start",
                    "end",
                    "CpG",
                    "ndx")],
            file = "Data/dt1.tsv",
            sep = "\t",
            row.names = FALSE)
```


# Annotate DNA samples
```{r}
peakAnno1 <- annotatePeak(peak = "data/dt1.tsv", #"data/combined_def.tsv", 
                          tssRegion = c(-3000, 3000), # default - CHECK!
                          TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene,
                          annoDb = "org.Mm.eg.db") # provides the following extra columns:
                                                   # ENSEMBL, SYMBOL, GENENAME
```

## Save annotated data
```{r}
peakAnno1@annoStat
dt2 <- data.table(as.data.frame(peakAnno1@anno@elementMetadata@listData))
setkey(dt2)
# dt2 <- dt2[dt2$SYMBOL != "NA", ]
head(dt2)

length(unique(dt2$SYMBOL))
unique(substr(x = dt2$annotation, 
              start = 1, 
              stop = 4))
names(dt2)[2] <- "pos"
```

# Package DSS
Dispersion shrinkage for sequencing data 

## Prepare DSS data list
```{r}
l_dt_dss <- list(data.table(chr = dt1$chr,
                            # pos = dt1$start,
                            pos = dt1$ndx,
                            N = dt1$`SL01_R1_001_val_1_bismark_bt2_pe-N`,
                            X = dt1$`SL01_R1_001_val_1_bismark_bt2_pe-X`),
                 data.table(chr = dt1$chr,
                            # pos = dt1$start,
                            pos = dt1$ndx,
                            N = dt1$`SL02_R1_001_val_1_bismark_bt2_pe-N`,
                            X = dt1$`SL02_R1_001_val_1_bismark_bt2_pe-X`),
                 data.table(chr = dt1$chr,
                            # pos = dt1$start,
                            pos = dt1$ndx,
                            N = dt1$`SL03_R1_001_val_1_bismark_bt2_pe-N`,
                            X = dt1$`SL03_R1_001_val_1_bismark_bt2_pe-X`),
                 data.table(chr = dt1$chr,
                            # pos = dt1$start,
                            pos = dt1$ndx,
                            N = dt1$`SL04_R1_001_val_1_bismark_bt2_pe-N`,
                            X = dt1$`SL04_R1_001_val_1_bismark_bt2_pe-X`),
                 data.table(chr = dt1$chr,
                            # pos = dt1$start,
                            pos = dt1$ndx,
                            N = dt1$`SL05_R1_001_val_1_bismark_bt2_pe-N`,
                            X = dt1$`SL05_R1_001_val_1_bismark_bt2_pe-X`),
                 data.table(chr = dt1$chr,
                            # pos = dt1$start,
                            pos = dt1$ndx,
                            N = dt1$`SL06_R1_001_val_1_bismark_bt2_pe-N`,
                            X = dt1$`SL06_R1_001_val_1_bismark_bt2_pe-X`),
                 data.table(chr = dt1$chr,
                            # pos = dt1$start,
                            pos = dt1$ndx,
                            N = dt1$`SL07_R1_001_val_1_bismark_bt2_pe-N`,
                            X = dt1$`SL07_R1_001_val_1_bismark_bt2_pe-X`),
                 data.table(chr = dt1$chr,
                            # pos = dt1$start,
                            pos = dt1$ndx,
                            N = dt1$`SL08_R1_001_val_1_bismark_bt2_pe-N`,
                            X = dt1$`SL08_R1_001_val_1_bismark_bt2_pe-X`))
```

## Make DSS object
```{r}
obj_dss <- makeBSseqData(dat = l_dt_dss,
                         sampleNames = c("WT_Control_1",
                                         "WT_Control_2",
                                         "WT_4HNE_1",
                                         "WT_4HNE_2",
                                         "KO_Control_1",
                                         "KO_Control_2",
                                         "KO_4HNE_1",
                                         "KO_4HNE_2"))
```

## Pairwise tests
```{r}
system.time(ctl <- DMLtest(BSobj = obj_dss,
                           group1 = c("WT_Control_1",
                                      "WT_Control_2"),
                           group2 = c("KO_Control_1",
                                      "KO_Control_2"),
                           equal.disp = FALSE))
save(ctl,
     file = "tmp/ctl.RData")
gc()

system.time(wt <- DMLtest(BSobj = obj_dss,
              group1 = c("WT_Control_1",
                         "WT_Control_2"),
              group2 = c("WT_4HNE_1",
                         "WT_4HNE_2"),
              equal.disp = FALSE))
save(wt,
     file = "tmp/wt.RData")
gc()

system.time(ko <- DMLtest(BSobj = obj_dss,
              group1 = c("KO_Control_1",
                         "KO_Control_2"),
              group2 = c("KO_4HNE_1",
                         "KO_4HNE_2"),
              equal.disp = FALSE))
save(ko,
     file = "tmp/ko.RData")
gc()
```

## Controls: WT vs. KO
### Reload test results
```{r}
load("tmp/ctl.RData")
```

### Merge with annotation and save
```{r}
ctl_anno <- merge(dt2,
                  ctl,
                  by = "pos")

tmp <- dt1[, c("ndx",
               "chr",
               "start",
               "end")]
names(tmp)[1] <- "pos"
ctl_anno <- merge(tmp,
                  ctl_anno,
                  by = "pos")

setorder(ctl_anno,
         fdr)
write.table(ctl_anno,
            file = "docs/controls.tsv",
            row.names = FALSE,
            sep = "\t")
```

## WT: 4HNE vs Control
### Reload test results
```{r}
load("tmp/wt.RData")
```

### Merge with annotation and save
```{r}
wt_anno <- merge(dt2,
                  wt,
                  by = "pos")

tmp <- dt1[, c("ndx",
               "chr",
               "start",
               "end")]
names(tmp)[1] <- "pos"
wt_anno <- merge(tmp,
                  wt_anno,
                  by = "pos")

setorder(wt_anno,
         fdr)
write.table(wt_anno,
            file = "docs/wt.tsv",
            row.names = FALSE,
            sep = "\t")
```

## KO: 4HNE vs Control
### Reload test results
```{r}
load("tmp/ko.RData")
```

### Merge with annotation and save
```{r}
ko_anno <- merge(dt2,
                  ko,
                  by = "pos")

tmp <- dt1[, c("ndx",
               "chr",
               "start",
               "end")]
names(tmp)[1] <- "pos"
ko_anno <- merge(tmp,
                  ko_anno,
                  by = "pos")

setorder(ko_anno,
         fdr)
write.table(ko_anno,
            file = "docs/ko.tsv",
            row.names = FALSE,
            sep = "\t")
```

## Visualization of DMRs
![](media/Structure_of_Eukaryotic_Gene_SMC.jpg)
### Reload Control data
```{r}
ctl <- fread("docs/controls.tsv")
ctl$reg <- substr(x = ctl$annotation,
                  start = 1,
                  stop = 4)
unique(ctl$reg)
ctl$reg <- factor(ctl$reg,
                  levels = c("Dist",
                             "Prom",
                             "5' U",
                             "Exon",
                             "Intr",
                             "3' U",
                             "Down"),
                  labels = c("Distal Intergenic",
                             "Promoter (<=3kb)",
                             "5' Untranslated Region (UTR)",
                             "Exon",
                             "Intron",
                             "3' Untranslated Region (UTR)",
                             "Downstream (<=300bp)"))
```

```{r}
unique(ctl$annotation[ctl$reg == "Dist"])
```


```{r}
tmp_l <- list(data.table(chr = dt1$chr,
                         # pos = dt1$start,
                         pos = dt1$ndx,
                         N = dt1$`SL01_R1_001_val_1_bismark_bt2_pe-N`,
                         X = dt1$`SL01_R1_001_val_1_bismark_bt2_pe-X`),
              data.table(chr = dt1$chr,
                         # pos = dt1$start,
                         pos = dt1$ndx,
                         N = dt1$`SL02_R1_001_val_1_bismark_bt2_pe-N`,
                         X = dt1$`SL02_R1_001_val_1_bismark_bt2_pe-X`),
              data.table(chr = dt1$chr,
                         # pos = dt1$start,
                         pos = dt1$ndx,
                         N = dt1$`SL03_R1_001_val_1_bismark_bt2_pe-N`,
                         X = dt1$`SL03_R1_001_val_1_bismark_bt2_pe-X`),
              data.table(chr = dt1$chr,
                         # pos = dt1$start,
                         pos = dt1$ndx,
                         N = dt1$`SL04_R1_001_val_1_bismark_bt2_pe-N`,
                         X = dt1$`SL04_R1_001_val_1_bismark_bt2_pe-X`))

tmp_obj <- makeBSseqData(dat = tmp_l,
                         sampleNames = c("WT_Control_1",
                                         "WT_Control_2",
                                         "WT_4HNE_1",
                                         "WT_4HNE_2"))

tmp <- callDML(DMLresult = ctl,
               delta = 0.1,
               p.threshold = 0.05)
head(tmp)
showOneDMR(OneDMR = tmp[1, ],
           BSobj = tmp_obj)
```


## ANOVA
### Study desing
```{r}
design <- data.table(Treatment = rep(c("WT - Control",
                                       "WT + 4HNE",
                                       "KO - Control",
                                       "KO + 4HNE"),
                                     each = 2))
design$Treatment <- factor(design$Treatment,
                           levels = unique(design$Treatment))
```

```{r}
DMLfit1 <- DMLfit.multiFactor(BSobj = dtBS,
                             design = design,
                             formula = ~ Treatment)
summary(DMLfit1)
```

```{r}
DMLfit1$X
head(DMLfit1$fit$beta)
colnames(DMLfit1$X)
```

```{r}
comp1 <- DMLtest.multiFactor(DMLfit1, 
                             coef = "TreatmentWT + 4HNE")
head(comp1)

comp2 <- DMLtest.multiFactor(DMLfit1, 
                             coef = "TreatmentKO - Control")
head(comp2)

comp3 <- DMLtest.multiFactor(DMLfit1, 
                             coef = "TreatmentKO + 4HNE")
head(comp3)
```


```{r}
# TxDb.Mmusculus.UCSC.mm10.knownGene
```

```{r}
# org.Mm.eg.db
# columns(org.Mm.eg.db)
```

```{r}
# dt1$chr <- factor(dt1$chr,
#                   levels = c(paste0("chr",
#                                     1:19),
#                              "chrX",
#                              "chrY",
#                              "chrM"))
# levels(dt1$chr)
```

```{r}
?annotatePeak
```

# Annotate DNA samples
```{r}
peakAnno1 <- annotatePeak(peak = "data/combined_def.tsv", 
                          tssRegion = c(-3000, 3000), # default - CHECK!
                          TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene,
                          annoDb = "org.Mm.eg.db") # provides the following extra columns:
                                                   # ENSEMBL, SYMBOL, GENENAME
```

## Save annotated data
```{r}
peakAnno1@annoStat
dt2 <- data.table(as.data.frame(peakAnno1@anno@elementMetadata@listData))
setkey(dt2)
# dt2 <- dt2[dt2$SYMBOL != "NA", ]
head(dt2)

length(unique(dt2$SYMBOL))
unique(substr(x = dt2$annotation, 
              start = 1, 
              stop = 4))
```

## Remove metochondrial DNA
```{r}
# unique(dt2$geneChr)
# dt2 <- dt2[geneChr != "chrM", ]
```

## Annotated CpGs (merge with p-values later)
```{r}
dt_anno <- data.table(chr = dt1$chr,
                      pos = dt2$geneStart,
                      end = dt2$geneEnd,
                      length = dt2$geneLength,
                      gene = dt2$SYMBOL,
                      dt2[, c("geneLength",
                              "geneStrand",
                              "distanceToTSS",
                              "CpG",
                              "annotation",
                              "geneId",
                              "transcriptId",
                              "ENSEMBL",
                              "GENENAME")])
head(dt_anno)
```

```{r}
peakAnnoM1 <- annotatePeak(peak = "tmp/m1.RData", 
                          tssRegion = c(-3000, 3000), # default - CHECK!
                          TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene,
                          annoDb = "org.Mm.eg.db")
```

### Merge with annotation
```{r}
m1$chr <- as.character(m1$chr)
m1 <- data.table(m1)
setkey(m1, 
       pos,
       chr)
unique(m1$chr)
m1$pos[1:10]
m1[chr == "chr1" &
     pos == 3038669, ]

dt_anno$chr <- as.character(dt_anno$chr)
setkey(dt_anno)
unique(dt_anno$chr)
dt_anno$pos[1:10]
dt_anno[chr == "chr1" &
     pos == 3038669, ]


dt1[chr == "chr1" &
     start == 3038669, ]
dtm1 <- merge(dt_anno,
              m1,
              by = c("chr",
                     "pos"))
head(dtm1)
```

# CONTINUE HERE (2/4/2022)!


```{r}
tmp <- dt2[, meta$sample_name, with = FALSE]
rs <- rowSums(tmp)
summary(rs)

row_keep <- which(!is.na(rs))
dt3 <- dt3[row_keep, ]
```

# Promoter CpGs only
```{r}
dt3[, reg := substring(text = annotation,
                       first = 1,
                       last = 4)]
proms <- dt3[reg == "Prom",]
```

```{r}
out <- list()
for (i in 13:18) {
  tmp <- proms[, c(11, i), with = FALSE]
  colnames(tmp)[2] <- "X"
  out[[i - 12]] <- tmp[, .(mu = mean(X)),
                       by = SYMBOL]
  colnames(out[[i - 12]])[2] <- paste0("X", i)
}
```

```{r}
res <- Reduce(function(...) merge(..., all = TRUE, by = "SYMBOL"), out)
```

```{r}
res[, Control := 100*X14/X13]
res[, LPS := 100*X16/X15]
res[, SAHA := 100*X18/X17]
res[, LPS_Control := LPS - Control]
res[, LPS_SAHA := LPS - SAHA]
res[, mu_LPS_Control := (LPS + Control)/2]
res[, mu_LPS_SAHA := (LPS + SAHA)/2]
```

# Control vs. LPS
```{r}
plot(res$LPS_Control ~ res$mu_LPS_Control,
     pch = ".")
```

## Average SD
```{r}
sd_LPS_Control <- sd(res$LPS_Control) 
sd_LPS_Control
```

## Predict SD based on mean methylation
```{r}
setorder(res,
         mu_LPS_Control)

# Move 1 unit (%) at the time
dt_ma <- data.table(mu_LPS_Control = 0:100)
dt_ma$mov_avg <- NA

# Window width = 10
for (i in 1:nrow(dt_ma)) {
  dt_ma$mov_avg[i] <- sd(res$LPS_Control[res$mu_LPS_Control >= dt_ma$mu_LPS_Control[i] &
                                res$mu_LPS_Control < dt_ma$mu_LPS_Control[i] + 10])
}

plot(dt_ma$mov_avg)

m1 <- loess(mov_avg ~ mu_LPS_Control,
            data = dt_ma,
            span = 0.2)
plot(predict(m1),
     type = "l")

res$pred_sd_LPS_Control <- predict(m1,
                       newdata = data.table(mu_LPS_Control = res$mu_LPS_Control))
```

# Statistics
```{r}
res[, stats_LPS_Control := LPS_Control/pred_sd_LPS_Control]
hist(res$stats_LPS_Control, 100)

plot(dnorm(res$stats_LPS_Control) ~ res$stats_LPS_Control,
     pch = ".")

m2 <- density(res$stats_LPS_Control,
              kernel = "gaussian",
              na.rm = TRUE)

m2
plot(m2,
     type = "p",
     pch = ".")

mean(m2$x)
sd(m2$x)
```


# LPS vs. SAHA
```{r}
plot(res$LPS_SAHA ~ res$mu_LPS_SAHA,
     pch = ".")
```

```{r}
dina <- meta$sample_name[meta$pi == "Dina"]
dna_dina <- dt2[, c("SYMBOL",
                    "distanceToTSS",
                    "annotation",
                    dina),
                with = FALSE]
head(dna_dina)
```

## Remove CpG clusters with small number of hits
### Set NAs to 0
```{r}
dna_dina <- data.table(dna_dina[, 1:3],
                       apply(X = dna_dina[, 4:ncol(dna_dina)],
                             MARGIN = 2,
                             FUN = function(a) {
                               a[is.na(a)] <- 0
                               return(a)
                             }))
```

### Total hits (X-es) per row
```{r}
dna_dina$row_hits <- apply(X = dna_dina[, seq(from = 5,
                                              to = ncol(dna_dina),
                                              by = 2),
                                        with = FALSE],
                           MARGIN = 1,
                           FUN = function(a) {
                             sum(a)
                           })

hist(row_hits, 100)
quantile(row_hits, c(0.5, 0.1))
ecdf(row_hits)(10)
```

### NOTE: almost 10% of rows have total hits (X) < 10. Remove the rows.
```{r}
dt1 <- dna_dina[row_hits >= 10,]

setorder(dt1,
         SYMBOL,
         distanceToTSS,
         annotation)
head(dt1)
rm(ls = list("beas_2b_methylseq_anno",
        "dt2",
        "dna_dina",
      ""))
rm(beas_2b_methylseq_anno)
rm(peakAnno1)
rm(dna_dina)
rm(dt2)
gc()
```
